### PR TITLE
[diffusion] Support LLaDA-Image BF16 generation and editing

### PR DESCRIPTION
**Pull request by rwang5203**
## Motivation

Add native SGLang Diffusion serving for [InclusionAI/LLaDA-Image](https://huggingface.co/InclusionAI/LLaDA-Image) and [InclusionAI/LLaDA-Image-Turbo](https://huggingface.co/InclusionAI/LLaDA-Image-Turbo), including their FP8 variants. The integration supports text-to-image generation and single-image editing through the OpenAI-compatible Images API, with single-GPU and sequence-parallel deployment.

## Contributors

- InclusionAI, Ant Group: Kun Chen (@kunchengit), Chengzhang Wu (@Anmuliar), Chaoyang Wang (@wzy-ustc)
- SGLang community: Kaige Li (@stephenkgli), Yuhan Zhou (@Dayuxiaoshui), Yihao Wang (@AgainstEntropy), Mick Qian (@mickqian), Richard Wang (@rwang5203)

## Modifications

- **Pipeline.** Add the LLaDA-Image DiT, pipeline configuration, sampling parameters, scheduler integration, embedded LLaDA2 MoE text encoder, source-image conditioning, and model registry entries.
- **Serving.** Support `/v1/images/generations` and `/v1/images/edits`, with request validation for image dimensions, pixel area, text length, output count, and exactly one edit source.
- **Sequence parallelism.** Support SP=1 and SP=2 with global stochastic-noise slicing, full text-encoder replication per SP rank, and a dedicated CFG FlashInfer attention path with fail-closed conditioning-mask validation.
- **FP8.** Support dynamic FP8 DiT weights and experts-only FP8 LLaDA2 checkpoints, including ModelOpt conversion, RMSNorm dtype handling, structured checkpoint metadata, legacy metadata compatibility, and component-path overrides.
- **Safety and lifecycle.** Propagate checkpoint revisions and remote-code trust into the embedded worker, reject unsupported residency and weight-update modes explicitly, and preserve generic non-conditioning attention behavior.

## Reproduction

Start Turbo with SP=1:

```bash
sglang serve --model-path inclusionAI/LLaDA-Image-Turbo --revision 033d068d0702ed81c45ec1bfbe61e318d843d948 --trust-remote-code --port 30000
```

Start Turbo with SP=2:

```bash
sglang serve --model-path inclusionAI/LLaDA-Image-Turbo --revision 033d068d0702ed81c45ec1bfbe61e318d843d948 --trust-remote-code --num-gpus 2 --sp-degree 2 --ulysses-degree 2 --ring-degree 1 --port 30000
```

Generate an image:

```bash
curl -sS -X POST http://127.0.0.1:30000/v1/images/generations -H "Content-Type: application/json" -d '{"prompt":"A cinematic photograph of a red fox standing in fresh snow at golden hour, detailed fur, shallow depth of field","size":"1024x1024","n":1,"response_format":"b64_json","num_inference_steps":4,"guidance_scale":1.0,"seed":42}'
```

Edit an image:

```bash
curl -sS -X POST http://127.0.0.1:30000/v1/images/edits -F "image=@input.png" -F "prompt=Place the subject on a beach" -F "size=1024x1024" -F "n=1" -F "response_format=b64_json" -F "num_inference_steps=4" -F "guidance_scale=1.0" -F "seed=42"
```

The production FP8 path uses the same commands with `inclusionAI/LLaDA-Image-Turbo-FP8` pinned to revision `2b822499be3f33a6f83c988dafab7986ebd1f733`.

## Accuracy Tests

The benchmark generated and decoded all 6704 expected artifacts: 2212 GenEval images, 1280 LongText images, 1212 GEdit images, and 2000 CVTG images.

| Configuration | GenEval macro | LongText en | LongText zh | CVTG Word Accuracy |
| --- | ---: | ---: | ---: | ---: |
| Base 50-step BF16 | 0.8404 | 0.9120 | 0.9198 | 0.7792 |
| Turbo 4-step SP=1 | 0.8190 | 0.8959 | 0.9123 | 0.8267 |
| Turbo 4-step SP=2 | 0.8190 | 0.8959 | 0.9123 | 0.8267 |

| Turbo GEdit metric | SP=1 | SP=2 |
| --- | ---: | ---: |
| en semantic / quality / overall | 7.9242 / 7.3682 / 7.5728 | 7.9175 / 7.3432 / 7.5451 |
| zh semantic / quality / overall | 7.7375 / 7.3867 / 7.4347 | 7.7327 / 7.4177 / 7.4534 |

GenEval, LongText, and CVTG reproduced the SP=1 scores exactly under SP=2. All six GEdit metrics remained within ±0.15 using the same evaluator.

Matched SP=1 and SP=2 T2I outputs were byte-identical. The matched edit output measured MAE 2.7935, PSNR 30.1759 dB, and RGB Pearson correlation 0.9919.

The production Turbo-FP8 checkpoint passed live SP=1 and SP=2 T2I, image editing, repeat determinism, and request-boundary validation at 1024x1024 with 4 steps and CFG 1.0.

The final post-rebase suite passed 59 tests and 4 subtests. The broader affected selection passed 341 tests and 102 subtests, including live CUDA RMSNorm coverage. The repository-configured Ruff, Ruff format, isort, and `git diff --check` checks also passed.

## Speed Tests and Profiling

The following values are median client-wall latency over ten sequential requests after two warmups on NVIDIA L20X at 1024x1024.

| Task | 4-step SP=1 | 4-step SP=2 | 50-step SP=1 | 50-step SP=2 |
| --- | ---: | ---: | ---: | ---: |
| T2I | 0.6216s | 0.4629s | 12.2036s | 7.6757s |
| Edit | 1.7717s | 1.2141s | 32.8138s | 19.9279s |

## Known Limitations

- The embedded text encoder requires reviewed remote code, so deployments must use `--trust-remote-code` and should pin the checkpoint revision.
- Sequence parallelism currently supports SP=1 or SP=2 with Ulysses only; SP=2 accepts one output per request.
- The embedded text encoder must remain resident and does not support generic memory sleep or hot weight updates.
- The full 6704-image quality suite was not rerun on the production FP8 checkpoint, so this PR does not claim FP8 quality-score parity.

## Checklist

- [x] Format the code according to the [SGLang pre-commit configuration](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [SGLang testing guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Provide deployment and reproduction instructions.
- [x] Provide accuracy and performance results.
- [x] Follow the [SGLang code-style guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #35693015041](https://github.com/sgl-project/sglang/actions/runs/35693015041)
Latest PR Test (Extra): :warning: **Not run on latest push** -- push again to dispatch.
Latest PR Test (AMD ROCm 10): :x: [Run #35693015036](https://github.com/sgl-project/sglang/actions/runs/35693015036)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/llada_image_conditioning.py
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/llada_image_conditioning.py:1]: move it to a dedicated folder under `model_specific_stages`

File: python/sglang/multimodal_gen/runtime/post_training/weights_updater.py [hunk 1]
```diff
@@ -367,6 +367,10 @@ def update_weights_from_disk(
         """Update model weights from disk without restarting the server."""
         logger.info(f"Updating weights from disk: {model_path}")
 
+        rejected, message = self._reject_if_partial_update()
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/post_training/weights_updater.py:370]: is rl support included in this pr?

File: python/sglang/srt/layers/attention/llada2_attention_utils.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (mickqian) [inline python/sglang/srt/layers/attention/llada2_attention_utils.py:1]: move all the content into `python/sglang/srt/layers/attention/llada2_cfg_flashinfer_backend.py`

File: python/sglang/srt/models/llada2_weight_utils.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# SPDX-License-Identifier: Apache-2.0
+"""Checkpoint-layout helpers for LLaDA2 models."""
```
Comment (mickqian) [inline python/sglang/srt/models/llada2_weight_utils.py:2]: util file should not be in this folder, please inline it or move elsewhere

File: python/sglang/multimodal_gen/runtime/loader/component_loaders/llada_image_loader.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/loader/component_loaders/llada_image_loader.py:1]: no need to duplicate a new loader, please refer to the recommended way in #38127

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/llada_image/conditioning.py
Comment (rwang5203) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/llada_image/conditioning.py:1]: Moved both conditioning stages into `model_specific_stages/llada_image` and updated the pipeline and test imports.

File: python/sglang/multimodal_gen/runtime/post_training/weights_updater.py (hunk 1 above)
Comment (rwang5203) [reply python/sglang/multimodal_gen/runtime/post_training/weights_updater.py:374]: No RL support is included in this PR. The guard rejects unsupported weight updates so the endpoint cannot partially update the LLaDA pipeline.

File: python/sglang/srt/layers/attention/llada2_attention_utils.py (hunk 2 above)
Comment (rwang5203) [reply python/sglang/srt/layers/attention/llada2_attention_utils.py:1]: The helper only depends on PyTorch, so its mask tests can run on CPU without FlashInfer. Could we keep it separate to preserve that test isolation?

File: python/sglang/srt/model_loader/llada2_weight_utils.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# SPDX-License-Identifier: Apache-2.0
+"""Checkpoint-layout helpers for LLaDA2 models."""
```
Comment (rwang5203) [reply python/sglang/srt/model_loader/llada2_weight_utils.py:2]: Moved the checkpoint helpers into `sglang.srt.model_loader.llada2_weight_utils` and updated the model and test imports.

File: python/sglang/multimodal_gen/runtime/loader/component_loaders/llada_image_loader.py (hunk 4 above)
Comment (rwang5203) [reply python/sglang/multimodal_gen/runtime/loader/component_loaders/llada_image_loader.py]: Switched QueryFormer, text projection, and SigVQ to the `component_loaders` mapping in #38127 and removed the dedicated loader. Checkpoint tests cover indexed shards, component and weight overrides, precision, and missing weights.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/llada_image/conditioning.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+            max_prefill_tokens=8192,
+            max_total_tokens=8192,
+            max_running_requests=4 if text_dp_attention else 2,
+            mem_fraction_static=(
+                server_args.pipeline_config.text_encoder_mem_fraction_static
+            ),
+        )
+        # TpModelWorker reads the runtime_context config bags the srt
+        # scheduler process publishes before construction.
+        srt_publish(srt_args, role="scheduler")
```
Comment (ClawSeven) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/llada_image/conditioning.py]: Could we avoid calling process-wide srt_publish() inside LLaDAImageTextEncoderRunner?

Comment (rwang5203) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/llada_image/conditioning.py]: Fixed. The encoder now uses its own runtime context for serialized initialization and encoding, restoring the diffusion context afterward.

**Review by mickqian — APPROVED** [review]

File: python/sglang/multimodal_gen/tools/build_modelopt_fp8_transformer.py [hunk 7]
```diff
@@ -850,6 +969,16 @@ def _parse_args() -> argparse.Namespace:
...
+    parser.add_argument(
+        "--activation-scheme",
+        choices=["auto", "static", "dynamic"],
+        default="auto",
+        help=(
+            "Activation quantization scheme in the converted checkpoint. 'auto' "
+            "uses dynamic activation quantization for LLaDA-Image and preserves "
+            "the legacy static ModelOpt format for other model families."
+        ),
+    )
```
Comment (AgainstEntropy) [inline python/sglang/multimodal_gen/tools/build_modelopt_fp8_transformer.py:981]: Can we parse the activation scheme from the model ckpt's ModelOpt config?

File: docs/src/snippets/diffusion/model-catalog.jsx
Comment (AgainstEntropy) [inline docs/src/snippets/diffusion/model-catalog.jsx:1]: Is cookbook coming in as separated PR?

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py [hunk 8]
```diff
@@ -0,0 +1,2502 @@
...
+from typing import ClassVar
+
+import torch
+import torch.nn.functional as F
+from diffusers.configuration_utils import ConfigMixin, register_to_config
+from diffusers.models.attention import AttentionMixin, AttentionModuleMixin, FeedForward
+from diffusers.models.attention_dispatch import dispatch_attention_fn
+from diffusers.models.modeling_outputs import Transformer2DModelOutput
+from diffusers.models.modeling_utils import ModelMixin
+from diffusers.models.normalization import RMSNorm
```
Comment (AgainstEntropy) [inline python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:26]: Can we use sglang native RMSNorm / RMSNormNoWeight?

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py [hunk 9]
```diff
@@ -0,0 +1,2502 @@
...
+        ) = self._batch_sequences(
+            list(image_features.split(image_lengths, dim=0)),
+            image_frequencies,
+            image_sequence.padding_masks,
+            self.x_pad_token,
+            image_sequence.noise_masks,
+        )
+
+        for layer in self.noise_refiner:
+            if torch.is_grad_enabled() and self.gradient_checkpointing:
```
Comment (AgainstEntropy) [inline python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:1261]: Since `self.gradient_checkpointing = False`, we can remove all gradient_checkpointing branches

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py
Comment (AgainstEntropy) [inline python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:1]: consider using `RowParallelLinear` / `ColumnParallelLinear` instead of `nn.Linear` across this file

File: docs/src/snippets/diffusion/model-catalog.jsx
Comment (Anmuliar) [reply docs/src/snippets/diffusion/model-catalog.jsx:1]: Cookbook has already been added in PR:https://github.com/Anmuliar/sglang/pull/2

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py (hunk 8 above)
Comment (Anmuliar) [reply python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:26]: To preserve model percision, SGLang's RMSNorm and Hugging Face's default RMSNorm differ in their casting behavior, so we use a mix of SGLang's and Diffusers' implementations.

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py
Comment (Anmuliar) [reply python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:1]: For the same precision-related reasons, the components that use nn.Linear account for less than 3% of the total runtime. We have replaced the linear layers in the main DiT backbone. If we can verify that this replacement does not affect the model’s overall performance, we can consider replacing these components as well.

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py (hunk 8 above)
Comment (Anmuliar) [reply python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:26]: To preserve model percision, SGLang's RMSNorm and Hugging Face's default RMSNorm differ in their casting behavior, so we use a mix of SGLang's and Diffusers' implementations.

File: docs/src/snippets/diffusion/model-catalog.jsx
Comment (Anmuliar) [reply docs/src/snippets/diffusion/model-catalog.jsx:1]: Cookbook has already been added in PR:https://github.com/Anmuliar/sglang/pull/2

Comment (rwang5203) [reply docs/src/snippets/diffusion/model-catalog.jsx:1]: Just merged the cookbook.

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py (hunk 8 above)
Comment (rwang5203) [reply python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:26]: The DiT blocks already use native RMSNorm with the reference cast order. H200 checks found small BF16 differences when replacing weightless Diffusers norms with RMSNormNoWeight, so I would retain them pending model-level parity.

File: python/sglang/multimodal_gen/tools/build_modelopt_fp8_transformer.py (hunk 7 above)
Comment (rwang5203) [reply python/sglang/multimodal_gen/tools/build_modelopt_fp8_transformer.py]: Done. Auto now reads the checkpoint’s activation metadata, with legacy defaults for missing fields and an explicit dynamic override for LLaDA-Image conversion.

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py (hunk 9 above)
Comment (rwang5203) [reply python/sglang/multimodal_gen/runtime/models/dits/llada_image.py]: Fixed. Removed the gradient-checkpointing branches and support flags from the DiT, QueryFormer, text projection, and SigVQ.

File: python/sglang/multimodal_gen/runtime/models/dits/llada_image.py
Comment (rwang5203) [reply python/sglang/multimodal_gen/runtime/models/dits/llada_image.py:1]: Makes sense. The main attention and FFN projections already use native parallel linears, and I suggest handling the remaining replicated projections in a focused follow-up with parity checks.

**Comment by alisonshao** [comment]
extra passed

**Comment by alisonshao** [comment]
baseline ci (all but h20 - in queue) passed

**Review by AgainstEntropy — APPROVED** [review]
